src/v/storage/disk_log_impl.cc

```diff
@@ -1529,7 +1529,26 @@ ss::future<> disk_log_impl::truncate_prefix(truncate_prefix_config cfg) {
     vassert(!_closed, "truncate_prefix() on closed log - {}", *this);
     return _failure_probes.truncate_prefix().then([this, cfg]() mutable {
         // dispatch the actual truncation
-        return do_truncate_prefix(cfg);
+        return do_truncate_prefix(cfg)
+          .then([this] {
```

Comment:
Just noting there's still room for a race where the reported size is lower than the reclaimable size, I think. Since size_bytes gets updated in `do_truncate()` and there is this scheduling point here.

It seems hard to be robust, so it's probably fine leaving this, but with the caveat that partition balancing should sanitize these values before making decisions.